<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Scripting for reproducible data analysis</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="css/style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Home</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="01-introduction.html">Introduction</a>
</li>
<li>
  <a href="00-start.html">Getting Started</a>
</li>
<li>
  <a href="02-filesystem.html">Working with the file system</a>
</li>
<li>
  <a href="03-files.html">Working with files</a>
</li>
<li>
  <a href="04-processes.html">Working with processes</a>
</li>
<li>
  <a href="05-analysis.html">Bioinformatics data analysis</a>
</li>
<li>
  <a href="06-scripts.html">Scripts</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Scripting for reproducible data
analysis</h1>

</div>


<head>
<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>
</head>
<hr />
<p>So far in this course we have learned how to work interactively on
the command line to analyse data. This section will introduce the idea
of <em>bash shell scripting</em>. We will also introduce the concept of
reproducible data analysis, and will show how using scripts to analyse
data can facilitate this.</p>
<p><br></p>
<div id="shell-scripts-in-bash" class="section level2">
<h2>Shell scripts in bash</h2>
<hr />
<br>
<div class="key-points">
<h2>
<i class="fas fa-thumbtack"></i> Key points
</h2>
<ul>
<li>A <strong>shell script</strong> is a text file containing multiple
commands, which can then be run from the command line as a single
command</li>
</ul>
</div>
<p><br></p>
<div id="writing-and-running-your-own-scripts" class="section level3">
<h3>Writing and running your own scripts</h3>
<hr />
<p>In its most basic form, a shell script is simply a text file
containing a list of commands that is run in sequence from top to
bottom. This kind of script can be run by providing it as an argument to
the <code>bash</code> command, as in the following example:</p>
<pre class="bash"><code>[USERNAME]@bifx-core2:~$ date
Thu 12 Nov 14:21:16 GMT 2020
[USERNAME]@bifx-core2:~$ echo hello
hello
[USERNAME]@bifx-core2:~$ ls /homes/genomes/mouse/
GRCm38  mm10  mm9  UCSC
[USERNAME]@bifx-core2:~$ echo date &gt; my_script.sh
[USERNAME]@bifx-core2:~$ echo &#39;echo hello&#39; &gt;&gt; my_script.sh
[USERNAME]@bifx-core2:~$ echo &#39;ls /homes/genomes/mouse&#39; &gt;&gt; my_script.sh
[USERNAME]@bifx-core2:~$ cat my_script.sh
date
echo hello
ls /homes/genomes/mouse/
[USERNAME]@bifx-core2:~$ bash my_script.sh
Thu 12 Nov 14:26:29 GMT 2020
hello
GRCm38  mm10  mm9  UCSC
[USERNAME]@bifx-core2:~$</code></pre>
<p>It is also possible to create a script that can be run as a
standalone program rather than as an argument to <code>bash</code>. This
involves two steps:</p>
<ol style="list-style-type: decimal">
<li>Add the line <code>#!/bin/bash</code> (known as a <a
href="https://en.wikipedia.org/wiki/Shebang_(Unix)">shebang</a> line) as
the first line of the file using a text editor</li>
<li>Use the <code>chmod</code> command to make the file executable
<ul>
<li><code>chmod a+x my_script.sh</code> will change the permissions on
the script so that any user can run it</li>
</ul></li>
</ol>
<p>Once these steps have been followed, it will be possible to execute
the script using its path:</p>
<pre class="bash"><code>[USERNAME]@bifx-core2:~$ cat my_script.sh
#!/bin/bash
date
echo hello
ls /homes/genomes/mouse/
[USERNAME]@bifx-core2:~$ ls -l ./my_script.sh
-rwxrwxr-x 1 [USERNAME] [USERNAME] 0 Nov 13 21:40 my_script.sh
[USERNAME]@bifx-core2:~$ ./my_script.sh
Thu 12 Nov 14:26:29 GMT 2020
hello
GRCm38  mm10  mm9  UCSC
[USERNAME]@bifx-core2:~$ </code></pre>
<p><em>Note:</em> If you want to be able to run your script just by
typing its name, you need to move it to a directory that is included in
the <em>$PATH</em> environment variable.</p>
<ul>
<li>To check which folders are listed in <em>$PATH</em>, you can type
<code>echo $PATH</code></li>
<li>To add a directory to <em>$PATH</em> permanently, add the line
<code>export PATH=[YOUR DIRECTORY]:$PATH</code> to the end of the
<code>~/.bashrc</code> file, then run the command
<code>source ~/.bashrc</code></li>
<li>If you run a program from a directory in <em>$PATH</em>, it can be
useful to check the full path to that program to make sure you’re not
inadvertently running another program with the same name. You can do
this by using the <code>which</code> command.</li>
</ul>
<p><br/></p>
</div>
<div id="editing-scripts" class="section level3">
<h3>Editing scripts</h3>
<hr />
<p>You can edit your scripts using a text editor such as <em>emacs</em>
or <em>vim</em>.</p>
<ul>
<li><code>emacs</code> opens emacs
<ul>
<li>you can then take the tutorial by typing Ctrl+h t</li>
</ul></li>
<li><code>vim</code> opens vim
<ul>
<li>you can take a tutorial by running the <code>vimtutor</code> command
at the bash prompt</li>
</ul></li>
</ul>
<p><br></p>
</div>
</div>
<div id="reproducible-data-analysis-with-scripts"
class="section level2">
<h2>Reproducible data analysis with scripts</h2>
<hr />
<br>
<div class="key-points">
<h2>
<i class="fas fa-thumbtack"></i> Key points
</h2>
<ul>
<li>Data analysis should be reproducible
<ul>
<li>You should be able to recreate all of the steps in your
analysis</li>
<li>Other researchers should also be able to recreate your analysis to
verify your results</li>
</ul></li>
<li>The bash shell makes it easy to analyse data interactively. However,
if data is analysed this way it can be difficult to keep track of
exactly which steps were taken to produce a given result</li>
<li>This problem can be solved by creating a script that contains all of
the commands needed to produce the results
<ul>
<li>You and other researchers can then recreate your analysis by running
the script</li>
</ul></li>
<li>This section presents a bash script that replicates the case study
performed in the previous section</li>
</ul>
</div>
<p><br></p>
<p>In the previous section we showed how to perform a simple analysis of
some RNA-Seq data from scratch. In order turn this into a reproducible
analysis workflow, it is necessary to put all of the commands that make
up the analysis into a script.</p>
<p><br></p>
<div id="creating-a-simple-data-analysis-script" class="section level3">
<h3>Creating a simple data analysis script</h3>
<hr />
<p>The simplest way to create a script is to put all of the commands
that made up the analysis into a file in the order in which they were
run, creating a file that looks like this:</p>
<pre class="bash"><code>mkdir analysis
cd analysis
mkdir 00_source_data
cd 00_source_data
ln -s ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq
cd ..
tree
mkdir 01_star_index
cd 01_star_index
ln -s ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta
nice STAR --runThreadN 5 --runMode genomeGenerate --genomeDir . --genomeFastaFiles ./yeast_genome.fasta --genomeSAindexNbases 10
cd ..
tree
mkdir 02_aligned_reads
cd 02_aligned_reads
nice STAR --genomeDir ../01_star_index/ --readFilesIn ../00_source_data/raw_yeast_rnaseq_data.fastq --runThreadN 5 --outFileNamePrefix raw_yeast_rnaseq_data. --outSAMtype BAM SortedByCoordinate
cd ..
tree
mkdir 03_coverage
cd 03_coverage
bedtools genomecov -ibam ../02_aligned_reads/raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam &gt; raw_yeast_rnaseq_data.genomecov.bg
cd ..
tree
mkdir 04_gene_overlap_counts
cd 04_gene_overlap_counts
ln -s ../../bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed
bedtools intersect -c -a yeast_genes.fixed.bed -b ../02_aligned_reads/raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam | awk -F&#39;\t&#39; &#39;$7&gt;0&#39; | sort -k7,7nr &gt; raw_yeast_overlap_data.gene_overlap_counts.bed
cd ..
tree
head -10 04_gene_overlap_counts/raw_yeast_overlap_data.gene_overlap_counts.bed
cd ..</code></pre>
<p>We can see that this works by running the above file (which is saved
on the server as
‘/library/training/bioinformatics_on_the_command_line/analysis_raw.sh’):</p>
<pre class="bash"><code>[USERNAME]@bifx-core2:~/course$ ls
analysis  bioinformatics_on_the_command_line_files  bioinformatics_on_the_command_line_files.tar.gz
[USERNAME]@bifx-core2:~/course$ mv analysis interactive_analysis
[USERNAME]@bifx-core2:~/course$ ls
bioinformatics_on_the_command_line_files  bioinformatics_on_the_command_line_files.tar.gz  interactive_analysis
[USERNAME]@bifx-core2:~/course$ bash /library/training/bioinformatics_on_the_command_line/analysis_raw.sh
.
└── 00_source_data
    └── raw_yeast_rnaseq_data.fastq -&gt; ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq

1 directory, 1 file
Nov 16 16:23:45 ..... started STAR run
Nov 16 16:23:45 ... starting to generate Genome files
Nov 16 16:23:46 ... starting to sort Suffix Array. This may take a long time...
Nov 16 16:23:46 ... sorting Suffix Array chunks and saving them to disk...
Nov 16 16:23:54 ... loading chunks from disk, packing SA...
Nov 16 16:23:55 ... finished generating suffix array
Nov 16 16:23:55 ... generating Suffix Array index
Nov 16 16:23:56 ... completed Suffix Array index
Nov 16 16:23:56 ... writing Genome to disk ...
Nov 16 16:23:56 ... writing Suffix Array to disk ...
Nov 16 16:23:57 ... writing SAindex to disk
Nov 16 16:23:57 ..... finished successfully
.
├── 00_source_data
│   └── raw_yeast_rnaseq_data.fastq -&gt; ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq
└── 01_star_index
    ├── chrLength.txt
    ├── chrNameLength.txt
    ├── chrName.txt
    ├── chrStart.txt
    ├── Genome
    ├── genomeParameters.txt
    ├── Log.out
    ├── SA
    ├── SAindex
    └── yeast_genome.fasta -&gt; ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta

2 directories, 11 files
Nov 16 16:23:57 ..... started STAR run
Nov 16 16:23:57 ..... loading genome
Nov 16 16:23:57 ..... started mapping
Nov 16 16:24:00 ..... finished mapping
Nov 16 16:24:00 ..... started sorting BAM
Nov 16 16:24:00 ..... finished successfully
.
├── 00_source_data
│   └── raw_yeast_rnaseq_data.fastq -&gt; ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq
├── 01_star_index
│   ├── chrLength.txt
│   ├── chrNameLength.txt
│   ├── chrName.txt
│   ├── chrStart.txt
│   ├── Genome
│   ├── genomeParameters.txt
│   ├── Log.out
│   ├── SA
│   ├── SAindex
│   └── yeast_genome.fasta -&gt; ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta
└── 02_aligned_reads
    ├── raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam
    ├── raw_yeast_rnaseq_data.Log.final.out
    ├── raw_yeast_rnaseq_data.Log.out
    ├── raw_yeast_rnaseq_data.Log.progress.out
    ├── raw_yeast_rnaseq_data.SJ.out.tab
    └── raw_yeast_rnaseq_data._STARtmp
        └── BAMsort
            ├── 0
            ├── 1
            ├── 2
            ├── 3
            └── 4

10 directories, 16 files
.
├── 00_source_data
│   └── raw_yeast_rnaseq_data.fastq -&gt; ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq
├── 01_star_index
│   ├── chrLength.txt
│   ├── chrNameLength.txt
│   ├── chrName.txt
│   ├── chrStart.txt
│   ├── Genome
│   ├── genomeParameters.txt
│   ├── Log.out
│   ├── SA
│   ├── SAindex
│   └── yeast_genome.fasta -&gt; ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta
├── 02_aligned_reads
│   ├── raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam
│   ├── raw_yeast_rnaseq_data.Log.final.out
│   ├── raw_yeast_rnaseq_data.Log.out
│   ├── raw_yeast_rnaseq_data.Log.progress.out
│   ├── raw_yeast_rnaseq_data.SJ.out.tab
│   └── raw_yeast_rnaseq_data._STARtmp
│       └── BAMsort
│           ├── 0
│           ├── 1
│           ├── 2
│           ├── 3
│           └── 4
└── 03_coverage
    └── raw_yeast_rnaseq_data.genomecov.bg

11 directories, 17 files
.
├── 00_source_data
│   └── raw_yeast_rnaseq_data.fastq -&gt; ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq
├── 01_star_index
│   ├── chrLength.txt
│   ├── chrNameLength.txt
│   ├── chrName.txt
│   ├── chrStart.txt
│   ├── Genome
│   ├── genomeParameters.txt
│   ├── Log.out
│   ├── SA
│   ├── SAindex
│   └── yeast_genome.fasta -&gt; ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta
├── 02_aligned_reads
│   ├── raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam
│   ├── raw_yeast_rnaseq_data.Log.final.out
│   ├── raw_yeast_rnaseq_data.Log.out
│   ├── raw_yeast_rnaseq_data.Log.progress.out
│   ├── raw_yeast_rnaseq_data.SJ.out.tab
│   └── raw_yeast_rnaseq_data._STARtmp
│       └── BAMsort
│           ├── 0
│           ├── 1
│           ├── 2
│           ├── 3
│           └── 4
├── 03_coverage
│   └── raw_yeast_rnaseq_data.genomecov.bg
└── 04_gene_overlap_counts
    ├── raw_yeast_overlap_data.gene_overlap_counts.bed
    └── yeast_genes.fixed.bed -&gt; ../../bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed

12 directories, 19 files
XII 460922  466869  RDN37-2 .   -   3730
XII 451785  457732  RDN37-1 .   -   3582
XII 468812  468931  RDN5-2  .   +   3063
XII 468826  468958  YLR154C-H   .   -   3063
XII 472464  472583  RDN5-3  .   +   3060
XII 472478  472610  YLR156C-A   .   -   3060
XII 482044  482163  RDN5-4  .   +   3055
XII 482058  482190  YLR157C-C   .   -   3055
XII 485696  485815  RDN5-5  .   +   3052
XII 485710  485842  YLR159C-A   .   -   3052

[USERNAME]@bifx-core2:~/course$ </code></pre>
<p><br></p>
</div>
<div id="creating-an-improved-data-analysis-script"
class="section level3">
<h3>Creating an improved data analysis script</h3>
<hr />
<p>In this section we improve our script by doing the following:</p>
<ul>
<li>Adding a shebang line to the script so that it can be run as a
standalone program</li>
<li>Ensuring that the script does not keep running if one of the
commands within it fails
<ul>
<li>This is what the <code>set -euo pipefail</code> command at the top
of the script does</li>
</ul></li>
<li>Deleting unnecessary commands</li>
<li>Adding spacing and comments to make the script easier to read
<ul>
<li>In bash, lines starting with <code>#</code> are taken to be
comments</li>
</ul></li>
<li>Specifying inputs as variables in order to make the script more
general and easier to maintain, and using paths relative to ‘~’ rather
than the current working directory so that the script can be run in any
directory and will still find the files</li>
<li>Adding a few lines to the script to create a run report, which
specifies the software versions used for the analysis</li>
</ul>
<p>The updated script is shown here:</p>
<pre class="bash"><code>#!/bin/bash
# This is simple RNA-Seq analysis workflow that does the following:
# - Creates a STAR index for the genome fasta file specified in the variable GENOME_FASTA
# - Aligns the raw reads specified in the fastq file referenced by the variable RAW_FASTQ
# - Computes the genome coverage of the aligned reads, producing a bedGraph file
# - Counts the overlaps over the genes specified in the file referenced by the variable GENES_BED

set -euo pipefail


# Global variables ----

RAW_FASTQ=&#39;~/course/bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq&#39;
GENOME_FASTA=&#39;~/course/bioinformatics_on_the_command_line_files/yeast_genome.fasta&#39;
GENES_BED=&#39;~/course/bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed&#39;

ls &quot;$RAW_FASTQ&quot; &quot;$GENOME_FASTA&quot; &quot;$GENES_BED&quot; &gt; /dev/null

RAW_FASTQ_BASENAME_PREFIX=`basename $RAW_FASTQ .fastq`
GENOME_FASTA_BASENAME=`basename $GENOME_FASTA`

THREADS=5


# Pipeline commands ----

# Create a directory for the results of the analysis

mkdir analysis
cd analysis

# Link to the fastq containing the raw sequences in &#39;00_source_data&#39;

mkdir 00_source_data
cd 00_source_data
ln -s &quot;$RAW_FASTQ&quot;
cd ..

# Create a STAR index for the genome fasta file, storing all of the results in &#39;01_star_index&#39;

echo &#39;Creating STAR index...&#39;

mkdir 01_star_index
cd 01_star_index
ln -s &quot;$GENOME_FASTA&quot;
nice STAR --runThreadN &quot;$THREADS&quot; --runMode genomeGenerate --genomeDir . --genomeFastaFiles &quot;$GENOME_FASTA_BASENAME&quot; --genomeSAindexNbases 10
cd ..

# Align the raw sequences using STAR, storing all of the results in &#39;02_aligned_reads&#39;

echo &#39;Aligning raw reads...&#39;

mkdir 02_aligned_reads
cd 02_aligned_reads
nice STAR --genomeDir ../01_star_index/ --readFilesIn ../00_source_data/&quot;$RAW_FASTQ_BASENAME_PREFIX&quot;.fastq --runThreadN &quot;$THREADS&quot; --outFileNamePrefix &quot;$RAW_FASTQ_BASENAME_PREFIX&quot;. --outSAMtype BAM SortedByCoordinate
cd ..

# Create a genome coverage file in bedGraph format, and store it in &#39;03_coverage&#39;

echo &#39;Creating genome coverage file...&#39;

mkdir 03_coverage
cd 03_coverage
bedtools genomecov -ibam ../02_aligned_reads/&quot;$RAW_FASTQ_BASENAME_PREFIX&quot;.Aligned.sortedByCoord.out.bam &gt; &quot;$RAW_FASTQ_BASENAME_PREFIX&quot;.genomecov.bg
cd ..

# Compute the gene overlap counts, and store them as a bed file in &#39;04_gene_overlap_counts&#39;

echo &#39;Computing gene overlap counts...&#39;

mkdir 04_gene_overlap_counts
cd 04_gene_overlap_counts
ln -s ../../bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed
bedtools intersect -c -a &quot;$GENES_BED&quot; -b ../02_aligned_reads/&quot;$RAW_FASTQ_BASENAME_PREFIX&quot;.Aligned.sortedByCoord.out.bam | awk -F&#39;\t&#39; &#39;$7&gt;0&#39; | sort -k7,7nr &gt; &quot;$RAW_FASTQ_BASENAME_PREFIX&quot;.gene_overlap_counts.bed
cd ..

# Create a run report

echo `realpath $0`&quot; run completed successfully.&quot; &gt; run_report.txt
date &gt;&gt; run_report.txt
echo &#39;Software versions:&#39; &gt;&gt; run_report.txt
echo &#39;STAR&#39; &gt;&gt; run_report.txt
STAR --version &gt;&gt; run_report.txt
echo &#39;bedtools&#39; &gt;&gt; run_report.txt
bedtools --version | sed &#39;s/^bedtools //&#39; &gt;&gt; run_report.txt
echo &#39;sort&#39; &gt;&gt; run_report.txt
sort --version | sed &#39;s/^sort //&#39; | head -1 &gt;&gt; run_report.txt
echo &#39;awk&#39; &gt;&gt; run_report.txt
awk --version | head -1 &gt;&gt; run_report.txt

# If we&#39;ve got here the pipeline has completed successfully

echo &#39;Pipeline completed successfully.&#39;</code></pre>
<p>Again, we can see that this works by running the above file (which is
saved on the server as
‘/library/training/bioinformatics_on_the_command_line/analysis_improved.sh’):</p>
<pre class="bash"><code>[USERNAME]@bifx-core2:~/course$ ls
analysis  bioinformatics_on_the_command_line_files  bioinformatics_on_the_command_line_files.tar.gz interactive_analysis
[USERNAME]@bifx-core2:~/course$ mv analysis analysis_raw
[USERNAME]@bifx-core2:~/course$ ls
analysis_raw                              bioinformatics_on_the_command_line_files.tar.gz
bioinformatics_on_the_command_line_files  interactive_analysis
[USERNAME]@bifx-core2:~/course$ bash /library/training/bioinformatics_on_the_command_line/analysis_improved.sh
Creating STAR index...
Nov 16 16:38:09 ..... started STAR run
Nov 16 16:38:09 ... starting to generate Genome files
Nov 16 16:38:10 ... starting to sort Suffix Array. This may take a long time...
Nov 16 16:38:10 ... sorting Suffix Array chunks and saving them to disk...
Nov 16 16:38:18 ... loading chunks from disk, packing SA...
Nov 16 16:38:19 ... finished generating suffix array
Nov 16 16:38:19 ... generating Suffix Array index
Nov 16 16:38:20 ... completed Suffix Array index
Nov 16 16:38:20 ... writing Genome to disk ...
Nov 16 16:38:20 ... writing Suffix Array to disk ...
Nov 16 16:38:21 ... writing SAindex to disk
Nov 16 16:38:21 ..... finished successfully
Aligning raw reads...
Nov 16 16:38:21 ..... started STAR run
Nov 16 16:38:21 ..... loading genome
Nov 16 16:38:21 ..... started mapping
Nov 16 16:38:24 ..... finished mapping
Nov 16 16:38:24 ..... started sorting BAM
Nov 16 16:38:24 ..... finished successfully
Creating genome coverage file...
Computing gene overlap counts...
Pipeline completed successfully.
[USERNAME]@bifx-core2:~/course$ cat analysis/run_report.txt
/library/training/bioinformatics_on_the_command_line/analysis_improved.sh run completed successfully.
Mon 16 Nov 16:38:36 GMT 2020
Software versions:
STAR
2.7.3a
bedtools
v2.27.0
sort
(GNU coreutils) 8.28
awk
GNU Awk 4.1.4, API: 1.1 (GNU MPFR 4.0.1, GNU MP 6.1.2)
[USERNAME]@bifx-core2:~/course$ </code></pre>
<br>
<div class="challenge">
<h2>
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
What is the line
<code>ls "$RAW_FASTQ" "$GENOME_FASTA" "$GENES_BED" &gt; /dev/null</code>
in the script for?
<details>
<summary>
</summary>
<div class="solution">
<h2>
<i class="far fa-eye"></i> Solution:
</h2>
<p>If the files referenced by <em>RAW_FASTQ</em>, <em>GENOME_FASTA</em>,
and <em>GENES_BED</em> exist, the line does nothing, as the output of
the command is redirected to <code>/dev/null</code> (which basically
means that it is ignored). However, if any of the files don’t exist this
command will fail with an error message. Because we have included the
line <code>set -eou pipefail</code> this will cause the script to stop
running, which is what we want if one or more of the output files is
missing.</p>
</div>
</details>
</div>
<p><br></p>
</div>
<div id="managing-your-scripts" class="section level3">
<h3>Managing your scripts</h3>
<hr />
<p>Once you start writing scripts, it is a good idea to use a
<em>version control system</em> to keep track of the changes you make to
your scripts. A good choice for this would be <code>git</code>. The
following five commands will allow you to use <code>git</code> for
version control:</p>
<ul>
<li><code>git init</code> to initialise a git repository in the current
working directory</li>
<li><code>git add -A; git commit -m "Latest updates"</code> to commit
any changes to files in the current working directory to the
repository</li>
<li><code>git status</code> to see if any changes have been made since
the last commit</li>
<li><code>git diff</code> to see the differences between the files in
the current working directory and the last commit</li>
<li><code>git show HEAD:[NAME OF FILE IN GIT]</code> to view the most
recent version of a file in the git repository
<ul>
<li>You can save the output of <code>git show</code> to a file by
redirecting its output using <code>&gt;</code></li>
</ul></li>
</ul>
<p>To learn more about <code>git</code>, you can look at this <a
href="https://swcarpentry.github.io/git-novice/">software carpentry
course</a>, which covers it in a lot more detail.</p>
<p><em>Note:</em> <code>git</code> should only be used for small text
files such as scripts. It is not designed to work with large data
files.</p>
<p><br></p>
</div>
<div id="closing-thoughts-practical-workflows-on-the-command-line"
class="section level3">
<h3>Closing thoughts: practical workflows on the command line</h3>
<hr />
<p>In this section we created a data analysis script in bash that
satisfies two major requirements for reproducible data analysis:</p>
<ul>
<li>The script provides a complete and accurate record of the steps that
were taken to produce the output files, along with a record of the
versions of the software tools used to manipulate the data</li>
<li>The script is also in a form that would be easy to share with other
researchers, allowing them to replicate your analysis easily</li>
</ul>
<p>The main drawback of using bash scripts to represent workflows is
that they are often impractical. The example presented here uses an
extremely small input FASTQ file, and an organism with a relatively
small genome. As a result it can be run from scratch in under a minute.
Running the same pipeline with a realistically sized input FASTQ file
and a larger genome could take hours to run, so it is no longer
practical to re-run the pipeline every time you make a change. A
workaround for this would be to comment out parts of the script that you
do not want to re-run, however this is considered bad practice as it
introduces the possibility of human error in selecting the parts of the
pipeline that need to be re-run when the pipeline is updated.</p>
<p>If you are writing analysis pipelines, you should look into using a
modern workflow system such as <a
href="https://snakemake.readthedocs.io/en/stable/">Snakemake</a> or <a
href="https://www.nextflow.io/">Nextflow</a>. Workflows written for
these workflow systems can be run repeatedly, and the workflow system
will automatically work out which steps of the analysis workflow need to
be re-run based on the timestamps of the files. They also provide a
number of other useful features:</p>
<ul>
<li>They work out which steps of the analysis workflow can be run in
parallel, and assign steps to different cores when possible, which can
speed up the analysis considerably</li>
<li>They simplify the process of running workflows on a computing
cluster</li>
</ul>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
